chall11/pcuadrad.c

```diff
@@ -0,0 +1,254 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <regex.h>
+
+#define isinset(a, b) (a == b) ? 1 : 0
+
+static int		count_str(char const *str, char c)
+{
+	int		count;
+
+	count = 0;
+	while (*str)
+	{
+		while (*str && isinset(*str, c))
+			str++;
+		if (*str && !isinset(*str, c))
+		{
+			count++;
+			while (*str && !isinset(*str, c))
+				str++;
+		}
+	}
+	return (count);
+}
+
+static char		*malloc_string(char *str, char c)
+{
+	char	*word;
+	int		i;
+
+	i = 0;
+	while (str[i] && !isinset(str[i], c))
+		i++;
+	if (!(word = (char *)malloc(sizeof(char) * (i + 1))))
+		return (NULL);
+	i = 0;
+	while (str[i] && !isinset(str[i], c))
+	{
+		word[i] = str[i];
+		i++;
+	}
+	word[i] = '\0';
+	return (word);
+}
+
+static char		**free_split(char ***tabstr)
+{
+	int i;
+
+	i = 0;
+	while (*tabstr[i])
+	{
+		free(*tabstr[i]);
+		i++;
+	}
+	free(*tabstr);
+	return (NULL);
+}
+
+char			**ft_split(char const *s, char c)
+{
+	int		len;
+	char	**tabstr;
+	int		i;
+
+	if (!s)
+		return (NULL);
+	len = count_str(s, c);
+	if (!(tabstr = (char **)malloc(sizeof(char*) * (len + 1))))
+		return (NULL);
+	i = 0;
+	while (*s)
+	{
+		while (*s && isinset(*s, c))
+			s++;
+		if (*s && !isinset(*s, c))
+		{
+			if (!(tabstr[i] = malloc_string((char*)s, c)))
+				return (free_split(&tabstr));
+			i++;
+			while (*s && !isinset(*s, c))
+				s++;
+		}
+	}
+	tabstr[i] = NULL;
+	return (tabstr);
+}
+
+char	*ft_strjoin(char const *s1, char const *s2)
+{
+	char	*joined;
+
+	if (!s1 && !s2)
+		return (NULL);
+	else if (s1 && !s2)
+		return (strdup(s1));
+	else if (!s1 && s2)
+		return (strdup(s2));
+	else
+	{
+		if (!(joined = (char*)malloc(sizeof(char) * (strlen(s1) + strlen(s2) + 2))))
+			return (NULL);
+        strcat(strcat(strcpy(joined, s1), " "), s2);
+		return (joined);
+	}
+}
+
+char    *parseSpaces(char *string)
+{
+    char    **temp;
+    char    *ret = NULL;
+    char    *tmp;
+    int     i;
+
+    temp = ft_split(string, ' ');
+    i = -1;
+    while (temp[++i])
+    {
+        if (!ret)
+        {
+            ret = strdup(temp[i]);
+            free(temp[i]);
+        }
+        else
+        {
+            tmp = ft_strjoin(ret, temp[i]);
+            free(ret);
+            free(temp[i]);
+            ret = tmp;
+        }
+    }
+    free(temp);
+    return (ret);
+}
+
+char    *isGotoStatement(char *str)
+{
+    regex_t regex;
+
+    if ((regcomp(&regex, "^goto", 0)))
+        return (NULL);
+    if (!(regexec(&regex, str, 0, NULL, 0)))
+    {
+        regfree(&regex);
+        return (strrchr(str, ' ') + 1);
+    }
+    regfree(&regex);
+    return (NULL);
+}
+
+int     infinity_loop(char **gotos, char *newGoto)
+{
+    int     i;
+
+    i = -1;
+    while(gotos[++i])
+        if (!strcmp(gotos[i], newGoto))
+            return (1);
+    return (0);
+}
+
+char    *treat_code(char **instr)
+{
+    char    *ret = NULL;
```

Comment:
There's no norm here, so it's actually more readable (personally), to group declarations/initializations like

```C
    char    *ret, *pattern, *tmp, **last_pattern = NULL;
    int     countGoto = 1;
    int     idx = 0;
```